src/ray/core_worker/reference_count.cc

```diff
@@ -1,28 +1,99 @@
+                 << " local_ref_count: " << it->second.local_ref_count                   \
+                 << " submitted_count: " << it->second.submitted_task_ref_count          \
+                 << " contained_in_owned: " << it->second.contained_in_owned.size()      \
+                 << " contained_in_borrowed: "                                           \
+                 << (it->second.contained_in_borrowed_id.has_value()                     \
+                         ? *it->second.contained_in_borrowed_id                          \
+                         : ObjectID::Nil())                                              \
+                 << " contains: " << it->second.contains.size();
+
+namespace {}  // namespace
+
 namespace ray {
 
-void ReferenceCounter::AddBorrowedObject(const ObjectID &object_id,
-                                         const TaskID &owner_id,
+ReferenceCounter::ReferenceTable ReferenceCounter::ReferenceTableFromProto(
+    const ReferenceTableProto &proto) {
+  ReferenceTable refs;
+  for (const auto &ref : proto) {
+    refs[ray::ObjectID::FromBinary(ref.reference().object_id())] =
+        Reference::FromProto(ref);
+  }
+  return refs;
+}
+
+void ReferenceCounter::ReferenceTableToProto(const ReferenceTable &table,
+                                             ReferenceTableProto *proto) {
+  for (const auto &id_ref : table) {
+    auto ref = proto->Add();
+    id_ref.second.ToProto(ref);
+    ref->mutable_reference()->set_object_id(id_ref.first.Binary());
+  }
+}
+
+bool ReferenceCounter::AddBorrowedObject(const ObjectID &object_id,
+                                         const ObjectID &outer_id, const TaskID &owner_id,
                                          const rpc::Address &owner_address) {
   absl::MutexLock lock(&mutex_);
+  return AddBorrowedObjectInternal(object_id, outer_id, owner_id, owner_address);
+}
+
+bool ReferenceCounter::AddBorrowedObjectInternal(const ObjectID &object_id,
+                                                 const ObjectID &outer_id,
+                                                 const TaskID &owner_id,
+                                                 const rpc::Address &owner_address) {
   auto it = object_id_refs_.find(object_id);
   RAY_CHECK(it != object_id_refs_.end());
 
-  if (!it->second.owner.has_value()) {
-    it->second.owner = {owner_id, owner_address};
+  RAY_LOG(DEBUG) << "Adding borrowed object " << object_id;
+  // Skip adding this object as a borrower if we already have ownership info.
+  // If we already have ownership info, then either we are the owner or someone
+  // else already knows that we are a borrower.
+  if (it->second.owner.has_value()) {
+    RAY_LOG(DEBUG) << "Skipping add borrowed object " << object_id;
+    return false;
   }
+
+  it->second.owner = {owner_id, owner_address};
+
+  if (!outer_id.IsNil()) {
+    auto outer_it = object_id_refs_.find(outer_id);
```

Comment:
How can this happen? Shouldn't we always also have a reference to the outer ID? Might be less error-prone to require that the outer ID is created first.